src/lib.rs

```diff
@@ -420,19 +418,65 @@ pub fn character(search_name: &str) -> Option<char> {
 
     // run through the parts of the name, matching them against the
     // parts of the input.
-    let mut passed_name = search_name;
+    let mut cmp_name = search_name;
     for part in maybe_name {
-        let part = part.as_bytes();
-        let part_l = part.len();
-        if passed_name.len() < part_l || &passed_name[..part_l] != part {
+        let part = match part {
+            "" => "-", // An empty word only appears before or after a non-medial hyphen
+            " " => continue,
+            "-" if codepoint != '\u{1180}' => continue,
+            part => part,
+        };
+
+        if let Some(rest) = cmp_name.strip_prefix(part.as_bytes()) {
+            cmp_name = rest;
+        } else {
             return character_by_alias(search_name);
         }
-        passed_name = &passed_name[part_l..]
+    }
+
+    // "HANGUL JUNGSEONG O-E" is ambiguous, returning U+116C HANGUL JUNGSEONG OE instead.
+    // All other ways of spelling U+1180 will get properly detected, so it's enough to just check
+    // if the hyphen is in the right place.
+    if codepoint == '\u{116C}'
+        && original_name
+            .trim_end_matches(|c: char| c.is_ascii_whitespace() || c == '_')
+            .bytes()
+            .nth_back(1)
+            == Some(b'-')
+    {
+        return Some('\u{1180}');
     }
 
     Some(codepoint)
 }
 
+fn normalise(search_name: &str, buf: &mut [u8]) -> usize {
+    let mut cursor = 0;
+    let bytes = search_name.as_bytes();
+
+    for (i, c) in bytes.iter().copied().enumerate() {
+        if c.is_ascii_whitespace() || c == b'_' {
+            continue;
+        }
+        if c == b'-'
+            && bytes.get(i - 1).map_or(false, u8::is_ascii_alphanumeric)
+            && bytes.get(i + 1).map_or(false, u8::is_ascii_alphanumeric)
+        {
+            continue;
+        }
+        if !c.is_ascii_alphanumeric() && c != b'-' {
+            return 0;
+        }
+        if cursor >= buf.len() {
+            return 0;
+        }
+        buf[cursor] = c.to_ascii_uppercase();
+        cursor += 1;
+    }
+
+    cursor
+}
```

Comment:
ditto.

Also add a comment on both implementations that they should match the other (except for HANGUL JUNGSEONG O-E; and explain that we don't check for HANGUL JUNGSEONG O-E here in order to avoid a string comparison that rarely matches).